rs/utils/src/bloom_filter/blocked_bloom_filter.rs

```diff
@@ -0,0 +1,115 @@
+    h1: u32,
+    h2: u32,
+}
+
+pub struct BlockedBloomFilter {
+    bits: BitVec<u8>,
+    num_blocks: usize,
+    num_hash_functions: usize,
+}
+
+impl BlockedBloomFilter {
+    pub const BLOCK_SIZE_IN_BITS: usize = 512;
+
+    pub fn new(expected_elements: usize, false_positive_prob: f64) -> Self {
+        // Calculate optimal parameters
+        let m = -((expected_elements as f64) * false_positive_prob.ln()) / (2.0_f64.ln().powi(2));
+        let m = m.ceil() as usize;
+
+        let k = (m as f64 / expected_elements as f64 * 2.0_f64.ln()).ceil() as usize;
+
+        // Round up to BLOCK_SIZE_IN_BITS blocks
+        let num_blocks = (m + Self::BLOCK_SIZE_IN_BITS - 1) / Self::BLOCK_SIZE_IN_BITS;
+
+        // Ensure bits is filled with false initially
+        let bits = bitvec![u8, Lsb0; 0; num_blocks * Self::BLOCK_SIZE_IN_BITS];
+
+        Self {
+            bits,
+            num_blocks,
+            num_hash_functions: k,
+        }
+    }
+
+    pub fn insert<T: Hash + ?Sized>(&mut self, key: &T) {
+        let hash_idx = self.hash_key(key);
+        let block_idx = self.get_block_idx(hash_idx.h1);
+        self.set_bits(hash_idx.h2, block_idx);
+    }
+
+    pub fn may_contain<T: Hash + ?Sized>(&self, key: &T) -> bool {
+        let hash_idx = self.hash_key(key);
+        let block_idx = self.get_block_idx(hash_idx.h1);
+        self.check_bits(hash_idx.h2, block_idx)
+    }
+
+    fn hash_key<T: Hash + ?Sized>(&self, key: &T) -> HashIdx {
+        let mut hasher = Xxh3::default();
+        key.hash(&mut hasher);
+        let hash = hasher.finish();
+        HashIdx {
+            h1: hash as u32,
+            h2: (hash >> 32) as u32,
+        }
+    }
+
+    fn get_block_idx(&self, h1: u32) -> usize {
+        // This is just FastRange32
+        (h1 as usize).wrapping_mul(self.num_blocks) >> 32
+    }
+
+    fn set_bits(&mut self, mut h: u32, block_idx: usize) {
+        let block_offset = block_idx * Self::BLOCK_SIZE_IN_BITS;
+        for _ in 0..self.num_hash_functions {
+            let bit_pos = (h >> (32 - Self::BLOCK_SIZE_IN_BITS.trailing_zeros())) as usize;
+            self.bits.set(block_offset + bit_pos, true);
+            h = h.wrapping_mul(0x9e3779b9);
+        }
+    }
+
+    fn check_bits(&self, mut h: u32, block_idx: usize) -> bool {
+        let block_offset = block_idx * Self::BLOCK_SIZE_IN_BITS;
+        for _ in 0..self.num_hash_functions {
+            let bit_pos = (h >> (32 - Self::BLOCK_SIZE_IN_BITS.trailing_zeros())) as usize;
+            if !self.bits[block_offset + bit_pos] {
+                return false;
+            }
+            h = h.wrapping_mul(0x9e3779b9);
+        }
+        true
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_blocked_bloom_filter() {
+        // Create a Bloom filter up to 4 billion documents and 1% false positive rate
+        let mut str_filter = BlockedBloomFilter::new(4_000_000_000, 0.01);
```

Comment:
let's make it smaller for test - test runs pretty slow if we allocate this much memory